### PR TITLE
Enable Shape Inference for NonZero operator

### PR DESCRIPTION
**Pull request by buddhapuneeth**
Currently Shape Inference is not there NonZero and based on the discussion [here](https://github.com/onnx/onnx/issues/3123).

### CODE REVIEW

**Comment by buddhapuneeth** [comment]
@jcwchen can you look into this PR? This is fix for the issue mentioned in the ticket https://github.com/onnx/onnx/issues/3123.

**Comment by jcwchen** [comment]
Hi @buddhapuneeth,
Thank you for proposing this PR. Could you add a simple test in `onnx/test/shape_inference_test.py`?
@onnx/sig-operators-approvers  please review this PR. Thanks!

**Review by postrational — APPROVED** [review]

**Comment by postrational** [comment]
Please add a test to `onnx/test/shape_inference_test.py` as @jcwchen mentioned. Otherwise this looks good to me.

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -2663,6 +2663,14 @@ ONNX_OPERATOR_SET_SCHEMA(
             "Constrain to all tensor types.")
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
           updateOutputElemType(ctx, 0, TensorProto::INT64);
+          auto* output_shape = ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+          auto& input_shape = getInputShape(ctx, 0);
+          int64_t rank = input_shape.dim_size();
+          // NonZero output is of rank 2.
+          // First dimension is same as input rank
+          output_shape->add_dim()->set_dim_value(rank);
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:2671]: I think that we should handle the case where the input rank is not known. Check ```if (hasInputShape(ctx,0))``` before calling ```getInputShape(ctx,0)```

Comment (gramalingam) [reply onnx/defs/tensor/defs.cc]: I made the suggested change. If someone could take a look, it would be helpful, thanks!

**Review by buddhapuneeth — COMMENTED** [review]
@gramalingam thanks for adding the check. The change looks good to me.

**Comment by gramalingam** [comment]
Closing this as this is covered by PR #3364

**Pull request closed without merging**